concent_api/core/payments/backends/sci_backend.py

```diff
@@ -32,31 +33,54 @@ def get_list_of_payments(
     """
     assert isinstance(requestor_eth_address,    str) and len(requestor_eth_address) == ETHEREUM_ADDRESS_LENGTH
     assert isinstance(provider_eth_address,     str) and len(provider_eth_address)  == ETHEREUM_ADDRESS_LENGTH
-    assert isinstance(payment_ts,               int) and payment_ts     >= 0
+    assert isinstance(min_block_timestamp, int) and min_block_timestamp >= 0
     assert isinstance(transaction_type,         Enum) and transaction_type in TransactionType
 
-    payment_interface: SmartContractsInterface = PaymentInterface()
+    payment_interface: SCIImplementation = PaymentInterface()
 
-    last_block_before_payment = BlocksHelper(payment_interface).get_first_block_after(payment_ts).number
+    first_block_after_payment_number = get_first_block_from_timestamp(payment_interface, min_block_timestamp).number
+    latest_block_number = payment_interface.get_block_number()
+    if latest_block_number - first_block_after_payment_number < payment_interface.REQUIRED_CONFS:
+        return []
 
     if transaction_type == TransactionType.FORCE:
-        payments_list = payment_interface.get_forced_payments(  # pylint: disable=no-member
-            requestor_address   = Web3.toChecksumAddress(requestor_eth_address),
-            provider_address    = Web3.toChecksumAddress(provider_eth_address),
-            from_block          = last_block_before_payment,
-            to_block            = payment_interface.get_block_number(),  # pylint: disable=no-member
+        payments_list = payment_interface.get_forced_payments(
+            requestor_address=Web3.toChecksumAddress(requestor_eth_address),
+            provider_address=Web3.toChecksumAddress(provider_eth_address),
+            from_block=first_block_after_payment_number,
+            to_block=latest_block_number - payment_interface.REQUIRED_CONFS,
         )
     elif transaction_type == TransactionType.BATCH:
-        payments_list = payment_interface.get_batch_transfers(  # pylint: disable=no-member
-            payer_address   = Web3.toChecksumAddress(requestor_eth_address),
-            payee_address   = Web3.toChecksumAddress(provider_eth_address),
-            from_block      = last_block_before_payment,
-            to_block        = payment_interface.get_block_number(),  # pylint: disable=no-member
+        payments_list = payment_interface.get_batch_transfers(
+            payer_address=Web3.toChecksumAddress(requestor_eth_address),
+            payee_address=Web3.toChecksumAddress(provider_eth_address),
+            from_block=first_block_after_payment_number,
+            to_block=latest_block_number - payment_interface.REQUIRED_CONFS,
         )
 
     return payments_list
 
 
+def get_first_block_from_timestamp(sci: SmartContractsInterface, timestamp: int) -> Block:
```

Comment:
This function is almost the same as in `BlocksHelper.get_first_block_after`, only marks `greater or equal` in 73 and 78 lines are different. Is it necessary?

Reply:
it is, these marks are everything ;) In the long run we could make a PR in `SCI` repo to move it there and extract some common code, but to fix problem in `concent`, what's done here is enough

Reply:
issue: https://github.com/golemfactory/concent/issues/1068

Reply:
Since timestamps are integers these two invocations are equivalent:
`get_first_block_from_timestamp(ts) == get_first_block_after(ts - 1)`
so this function is not necessary at all (nor any changes to SCI)